### DISCUSSION TITLE
Setting up custom property and reaction packages

### DISCUSSION THREAD

**Discussion by tk3016** [IDAES/idaes-pse#1338] [Q&A] [answered]
Hi,
I am a new user of IDAES and I am very excited to be using this amazing framework.
My aim is to model a catalytic reactor by building my custom model for methanol synthesis reaction and currently, I am trying to figure out how I can set up custom thermophysical and reaction property packages for the system I am interested in. My question is related to both property package and reaction property package. 

- **Thermophysical property package** -  I managed to set up my thermophysical property package and running the quick demo based on the example provided works. In my package, I am trying to define partial pressure (_pressure_comp_), a custom new variable that needs to be calculated based on the example provided to set up the package but I am not sure if my implementation conforms to the IDAES standards. As far as I understand, the build() method in the StateBlockData class is used exclusively for state variables. For additional calculations, one is supposed to define a separate method based on the name provided in the properties metadata and evaluate the expression, however this did not seem to work for me when I tried to calculate partial pressure of the components. It is currently being calculated in the build() method and any advice on the good practice here is very much appreciated. Please see the relevant screenshots of the code snippets below:
![image](https://github.com/IDAES/idaes-pse/assets/56822481/e205c17b-09ce-443b-ba09-9201018972fa)
![image](https://github.com/IDAES/idaes-pse/assets/56822481/7b3c5c0f-4c17-474f-9e8b-093e38a30567)

- **Reaction property package** - I am a bit lost here. I want to model the kinetics for methanol synthesis reaction as provided in this [paper](https://doi.org/10.1006/jcat.1996.0156). The closest example I could find on idaes is for [steam methane reforming](https://idaes-examples.readthedocs.io/en/stable/docs/unit_models/reactors/equilibrium_reactor_doc.html). However, as provided in the [paper](https://doi.org/10.1006/jcat.1996.0156), the rate equations are more complicated than using a simple power law expression, which if I understand correctly can be passed through a config dictionary. Again, any pointers would be very much appreciated.

Thank you very much in advance for all your support.

Best wishes,
Tanuj

**Comment by andrewlee94** [comment]
@Tanuj Welcome to IDAES.

To answer your questions:

1. You can put as many properties in the `build()` method as you want, not just the state variables - the reason why some properties have their own method is so that we can build them on when required (a lot of properties are only needed in specific unit models). Putting a property in the `build` method just means it will be build all the time (which is conceptually simpler) but can result in larger models with additional variables and constraints that are not required for what you are modeling. Ultimately, the choice is up to you. If you do put a property in the min `build` method, the "method" argument in the metadata should be `None` to indicate that this property should have been build by the `build` method (so that if something goes wrong it can be caught).
2. For your question about reaction packages, it appears you are looking at an example that uses `GenericReactionBlock` - this is a tool we developed to make it easy to quickly develop reaction packages for simple systems, but it does not support more complex reactions. You can write your own reaction packages however - a simple example is here: https://github.com/IDAES/idaes-pse/blob/main/idaes/models/properties/examples/saponification_reactions.py. This lets you specify any rate form you desire (although be careful as more complex rate forms can be poorly posed and result in potential solver issues).

**Reply by tk3016** [reply]
Many thanks for your prompt response Andrew, it is much clearer now. :)
Sorry to bother you (if I may) with a follow-up question related to setting up the reaction property package.
In the paper I referred to earlier, there are two equilibrium reactions with their corresponding rate constants, as shown in the screenshot below:
![image](https://github.com/IDAES/idaes-pse/assets/56822481/29558941-57dd-4494-88a7-0a30a397ac38)

I am trying to set up the reactions based on the example you provided by first creating a dictionary of two equilibrium reactions. As you can see in the screeenshot above, k_{MeOH} and k_{RWGS} have both Arrenhius style expressions, and I am trying to pass a dictionary as you can see below:
![image](https://github.com/IDAES/idaes-pse/assets/56822481/4821b9e8-5139-4165-a03b-05c843947241)

But on trying to run the script, I get the following error:
![image](https://github.com/IDAES/idaes-pse/assets/56822481/2d144d12-e547-49d3-9f7b-5502f06519f5)

Is it even possible to specify different Arrhenius constants for different reactions at once or am I missing something?

Thanks in advance again.

Best wishes,
Tanuj

**Comment by andrewlee94** [comment, marked as answer]
Hi again,

The first issue is that a single Pyomo component (e.g. `Param` is this case) can only have a single set of units; you cannot give different units for different indices. So, in your case you will need separate `Params` for each reaction. For the (relatively rare) case where all the Arrhenius constants have the same units, you can use a single indexed `Param`, but unfortunately Arrhenius constants tend to have weird units.

I think this is the source of your problem, although the error does not make this very clear (it is strictly correct, but does nothing to explain why it happened).

**Reply by tk3016** [reply]
Many thanks for your prompt response again, I have got my property package to work now.
